### ISSUE TITLE
What are the effective units of min and max budget?

### ISSUE THREAD

**Issue by Nicholas-Mitchell** [automl/HpBandSter#79]
I can't see anywhere in the documentation exactly what min and max budget mean. At first I saw examples that made it look like they are the number of epochs, but given the default values are 0.01 and 1, respectively, that doesn't make much sense.

Are they arbitrary, but it is simply the ratio of the two that determines the step sizes e.g. during successive halving?

**Comment by benevedes** [comment]
I think that the meaning of the budget is dependent on the problem. For tuning hyperparameters on a neural network, I think one would typically use it as number of epochs, e.g. see [the model.fit in the compute method in the worker for example 5.](https://automl.github.io/HpBandSter/build/html/auto_examples/example_5_keras_worker.html) However, in general, to the best of my understanding, they just parameterize the approximation of the objective function in some way that f(min_budget) is fast to evaluate and f(max_budget) is very close to the objective; you use the budgets in whatever way achieves that goal.

**Comment by Nicholas-Mitchell** [comment]
Thanks - I had figured it out in the end. There really aren't any
restrictions and it should be simply something meaningful in terms of the
available computational/time budget for the particular experiment. The only
tricky part is sometimes defining `f(min_budget) ` such that the loss finds
a reasonable trajectory.

After implementing a full distributed pipeline around hpbandster, I came to
appreciate how it can accommodate so much exploration!

**Comment by Ashwin-Ramesh2607** [comment]
@Nicholas-Mitchell So if I am training a model and it usually converges and produces a decent accuracy at around 300 epochs, what should I ideally keep my min and max budget as?

**Comment by Nicholas-Mitchell** [comment]
It depends on how long you want to wait, how long 1 epoch takes and how
much compute you are willing to spend.

Given 300 epochs for a full training, I'd probably do something like 3 as
minimum and 200-250 as maximum; wherever you are reasonably confident that
the loss won't make any big changes (i.e. it just slowly improves over the
last 50 epochs).

Making use of the number of iterations (number of successive halvings) -
and other settings like the best-k % of configurations that are allowed to
continue - you can control the distribution of your budget over the min-max
range.

It can be rather complicated because it depends on many factors, so might
be best to implement the whole pipeline and use a dummy model (or an empty
loop), so you can get a feeling for how to tweak the budget to your liking.

Good luck!

**Comment by Ashwin-Ramesh2607** [comment]
@Nicholas-Mitchell That was good, thanks. I am new to bayesian optimization and tuning techniques. Do you happen to know any resources or blogs that can get me upto speed on understanding how to better build this pipeline?

**Comment by Nicholas-Mitchell** [comment]
You're welcome!

Best source for actually building this pipeline is the hpbandster
documentation... Or if I were to write blog;)

To better understand the process of optimisation, I'd suggest the actual
BOHB Hpbandster paper (available on the Automl research group's website)
and it's citations. Don't forget that hyperband and the Bayesian
optimisation are two separate steps to consider. You could use them
independently too - there is an existing implementation of hyperband.
